File: Python/algo_codes/max_subarray_sum.py
# dc
def max_subarray_sum_DC(A: list) -> int:
    n = len(A)
    def max_crossing_sum(arr, l, m, h):
        # 找到包含中间元素的子数组的最大和（向左向右查找）
        sm = 0
        left_sum = float('-inf')
        for i in range(m, l-1, -1):
            sm += arr[i]
            if sm > left_sum:
                left_sum = sm
        sm = 0
        right_sum = float('-inf')
        for i in range(m+1, h+1):
            sm += arr[i]
            if sm > right_sum:
                right_sum = sm
        return left_sum + right_sum
    def max_subarray_sum(arr, l, h):
        # 递归计算最大子段和
        if l == h:  # 只有一个元素的情况
            return arr[l]
        m = (l + h) // 2
        # 分别找到左子数组、右子数组和包含中间元素的子数组的最大子段和
        left_sum = max_subarray_sum(arr, l, m)
        right_sum = max_subarray_sum(arr, m + 1, h)
        cross_sum = max_crossing_sum(arr, l, m, h)
        return max(left_sum, right_sum, cross_sum)


# dp
def max_subarray_sum_DP(A: list) -> int:
    n = len(A)
    max_so_far = A[0]
    curr_max = A[0]

    for i in range(1, n):
        curr_max = max(A[i], curr_max + A[i])
        max_so_far = max(max_so_far, curr_max)

    return max_so_far


# 测试代码
if __name__ == '__main__':
    arr = [-2, -5, 6, -2, -3, 1, 5, -6]
    max_sum = max_subarray_sum_DC(arr)
    print('最大子段和为：', max_sum)
    arr = [-2, -5, 6, -2, -3, 1, 5, -6]
    max_sum = max_subarray_sum_DP(arr)
    print('最大子段和为：', max_sum)

File: Cpp/algo_codes/max_subarray_sum.cpp
#include <iostream>
// 分治
int max(int a, int b, int c) {
  return a > b ? (a > c ? a : c) : (b > c ? b : c);
}

int maxCrossingSubarraySum(int A[], int low, int mid, int high) {
  int left_sum = 0;
  int right_sum = 0;
  int sum = 0;

  for (int i = mid; i >= low; i--) {
    sum += A[i];
    if (sum > left_sum) left_sum = sum;
  }

  sum = 0;
  for (int i = mid + 1; i <= high; i++) {
    sum += A[i];
    if (sum > right_sum) right_sum = sum;
  }

  return left_sum + right_sum;
}


int max_subarray_sum(int A[], int low, int high) {
  if (low == high) return A[low];

  int mid = (low + high) / 2;

  int left_sum = max_subarray_sum(A, low, mid);  // 递归计算左子序列最大子段和
  int right_sum = max_subarray_sum(A, mid + 1, high);  // 递归计算右子序列最大子段和
  int cross_sum =
      maxCrossingSubarraySum(A, low, mid, high);  // 跨越中间位置的最大子段和

  return max(left_sum, right_sum, cross_sum);  // 返回三者中的最大值
}

// dp
int max(int a, int b) {
  return a > b ? a : b;
}
int max_subarray_sum(int list[], int n) {
  int dp[n + 1];
  dp[0] = 0;
  int res = list[0];
  for (int i = 1; i < n + 1; i++) {
    dp[i] = max(dp[i - 1] + list[i - 1], dp[i - 1]);
    res = max(dp[i], res);
  }
  return res;
}


int main(){
  
  return 0;
}